cbmc/proofs/proof_guide.md

```diff
@@ -0,0 +1,399 @@
+[//]: # (SPDX-License-Identifier: CC-BY-4.0)
+
+# CBMC Proof Guide and Cookbook for MLKEM-C
+
+This doc acts as a guide to developing proofs of C code using CBMC. It concentrates
+on the use of _contracts_ to achieve _unbounded_ and _modular_ proofs of type-safety
+and correctness properties.
+
+This document uses the abbreviated forms of the CBMC contracts defined by macros in the
+cbmc.h header file in the MLKEM-C sources.
+
+## Common Proof Patterns
+
+### Loops (general advice)
+
+1. A function should contain at most one outermost loop statement. If the function you're trying to prove has more than one outermost loop, then re-factor it into two or more functions.
+
+2. The one outermost loop statement should be _final_ statement before the function returns. Don't have complicated code _after_ the loop body.
+
+### For loops
+
+The most common, and easiest, patten is a "for" loop that has a counter starting at 0, and counting up to some upper bound, like this:
+
+```
+for (int i = 0; i < C; i++) {
+    S;
+}
+```
+Notes:
+1. It is good practice to declare the counter variable locally, within the scope of the loop.
+2. The counter variable should be a constant within the loop body. In the example above, DO NOT modify `i` in the body of the loop.
+3. "int" is the best type for the counter variable. "unsigned" integer types complicate matters with their modulo N semantics. Avoid "size_t" since its large range (possibly unsigned 64 bit) can slow proofs down.
+
+CBMC requires basic assigns, loop-invariant, and decreases contracts _in exactly that order_. Note that the contracts appear _before_ the opening `{` of the loop body, so we also need to tell `clang-format` NOT to reformat these lines. The basic pattern is thus:
+
+```
+for (int i = 0; i < C; i++)
+    // clang-format off
+ASSIGNS(i) // plus whatever else S does
+INVARIANT(i >= 0)
+INVARIANT(i <= C)
+DECREASES(C - i)
+// clang-format on
+{
+    S;
+}
+```
+
+The `i <= C` in the invariant is NOT a typo. CBMC places the invariant just _after_ the loop counter has been incremented, but just _before_ the loop exit test, so it is possible for `i == C` at the invariant on the final iteration of the loop.
```

Comment:
I still find this confusing. Why is `INVARIANT(i >= 0)` not `INVARIANT(i > 0)`, then? 

In my mind, the invariant captures what remains true while the loop runs -- and that, in this instance, is `i >= 0` and `i < C`. So I share the confusion that you try to resolve here :-)

Reply:
For whatever reason, CBMC models such for loops:
```
for (int i = 0; i < C; i++)
INVARIANT(P)
{
  S;
}
```
as
```
int i = 0;
do
{
  INVARIANT(P);
  if !(i < C) break;
  S;
  i++;
} while(1);
```
Putting the exit test _after_ the Invariant means that it _is_ possible for i == C just before the loop exits.

Reply:
Thanks Rod, that's helpful

Reply:
You may want to add this here, as I won't be the only one confused by this.